<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2019-02-14" />

<title>DU-Bii Study cases - TCGA</title>

<script src="import_TCGA_from_Recount_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="import_TCGA_from_Recount_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="import_TCGA_from_Recount_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="import_TCGA_from_Recount_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="import_TCGA_from_Recount_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="import_TCGA_from_Recount_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="import_TCGA_from_Recount_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="import_TCGA_from_Recount_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="import_TCGA_from_Recount_files/navigation-1.1/tabsets.js"></script>
<script src="import_TCGA_from_Recount_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #303030; color: #cccccc; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #cccccc; background-color: #303030; }
code > span.kw { color: #f0dfaf; } /* Keyword */
code > span.dt { color: #dfdfbf; } /* DataType */
code > span.dv { color: #dcdccc; } /* DecVal */
code > span.bn { color: #dca3a3; } /* BaseN */
code > span.fl { color: #c0bed1; } /* Float */
code > span.ch { color: #dca3a3; } /* Char */
code > span.st { color: #cc9393; } /* String */
code > span.co { color: #7f9f7f; } /* Comment */
code > span.ot { color: #efef8f; } /* Other */
code > span.al { color: #ffcfaf; } /* Alert */
code > span.fu { color: #efef8f; } /* Function */
code > span.er { color: #c3bf9f; } /* Error */
code > span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */
code > span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code > span.sc { color: #dca3a3; } /* SpecialChar */
code > span.vs { color: #cc9393; } /* VerbatimString */
code > span.ss { color: #cc9393; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #f0dfaf; } /* ControlFlow */
code > span.op { color: #f0efd0; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #7f9f7f; } /* Documentation */
code > span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code > span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code > span.in { color: #7f9f7f; font-weight: bold; } /* Information */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DU-Bii Study cases - TCGA</h1>
<h3 class="subtitle"><em>DUBii 2019</em></h3>
<h4 class="author"><em>Jacques van Helden</em></h4>
<h4 class="date"><em>2019-02-14</em></h4>

</div>


<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Define parameters for the analysis
parameters &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">epsilon =</span> <span class="fl">0.1</span>,
  <span class="dt">gene.filter.min.count =</span> <span class="dv">10</span>,
  <span class="dt">gene.filter.percent.zeros =</span> <span class="dv">95</span>,
  <span class="dt">gzip.output.files =</span> <span class="ot">TRUE</span>
)

<span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">data.frame</span>(parameters)), <span class="dt">col.names =</span> <span class="st">&quot;Parameter value&quot;</span>)</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Parameter value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
epsilon
</td>
<td style="text-align:right;">
0.1
</td>
</tr>
<tr>
<td style="text-align:left;">
gene.filter.min.count
</td>
<td style="text-align:right;">
10.0
</td>
</tr>
<tr>
<td style="text-align:left;">
gene.filter.percent.zeros
</td>
<td style="text-align:right;">
95.0
</td>
</tr>
<tr>
<td style="text-align:left;">
gzip.output.files
</td>
<td style="text-align:right;">
1.0
</td>
</tr>
</tbody>
</table>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
</div>
<div id="data-source" class="section level2">
<h2>Data source</h2>
<p><strong>Recount2</strong> (<a href="https://jhubiostatistics.shinyapps.io/recount/" class="uri">https://jhubiostatistics.shinyapps.io/recount/</a>) is a database rasembling several thousands of human RNA-seq studies, that were all processed with a same workflow in order to ensure the consistency of the transcriptome measurements. Recount2 provides direct access to tables of raw counts per gene, exon or transcript.</p>
</div>
<div id="requirements" class="section level2">
<h2>Requirements</h2>
<ul>
<li>Recount library: <a href="https://bioconductor.org/packages/release/bioc/html/recount.html" class="uri">https://bioconductor.org/packages/release/bioc/html/recount.html</a></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Install BiocManager if required</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)

<span class="co"># Install recount if required</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">require</span>(recount)) {
  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;recount&quot;</span>, <span class="dt">version =</span> <span class="st">&quot;3.8&quot;</span>)
  <span class="kw">require</span>(recount)
}

<span class="co"># BiocManager::valid()</span></code></pre></div>
</div>
<div id="working-directory" class="section level2">
<h2>Working directory</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Set the working directory
workdir &lt;-<span class="st"> &quot;~/TCGA_import&quot;</span>
<span class="kw">dir.create</span>(workdir, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)
<span class="kw">setwd</span>(workdir)

## Select a TCGA project
selected.project &lt;-<span class="st"> &quot;Breast Invasive Carcinoma&quot;</span>
project.acronym &lt;-<span class="st"> &quot;BIC&quot;</span>
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Selected TCGA project</span><span class="ch">\t</span><span class="st">&quot;</span>, selected.project, <span class="st">&quot; (&quot;</span>, project.acronym, <span class="st">&quot;)&quot;</span>)

export.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;data&quot;</span>, project.acronym)
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Export directory</span><span class="ch">\t</span><span class="st">&quot;</span>, export.dir)
<span class="kw">dir.create</span>(export.dir, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>The downloaded and processed data will be saved in a working directory named <strong>TCGA_import</strong> in the user home folder (``r workdir). If required this directory is created automatically.</p>
</div>
<div id="data-source-1" class="section level2">
<h2>Data source</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">#### Download the TCGA data from Recount2 database ####

## Set the recoutnID (to &quot;TCGA&quot;)
recountID &lt;-<span class="st"> &quot;TCGA&quot;</span>

<span class="co"># Specify the download directory</span>
download.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(workdir, <span class="st">&quot;downloads&quot;</span>, recountID)
<span class="kw">dir.create</span>(download.dir, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)


## Data types to download
<span class="co">#download.types &lt;- c(&quot;phenotype&quot;, &quot;counts-gene&quot;, &quot;rse-gene&quot;)</span>
download.types &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;rse-gene&quot;</span>) <span class="co"># the rse-gene object contains both expression and pheno tables</span>
localFiles &lt;-<span class="st"> </span><span class="kw">c</span>()
<span class="cf">for</span> (type <span class="cf">in</span> download.types) {
  <span class="kw">message</span>(<span class="st">&quot;Recount data type: &quot;</span>, type)
  ## Get the URL of the recount file 
  ## (its extension will depend on the data type)
  recountURL &lt;-<span class="st"> </span><span class="kw">download_study</span>(
    recountID, 
    <span class="dt">outdir =</span> download.dir, 
    <span class="dt">type =</span> type, 
    <span class="dt">download =</span> <span class="ot">FALSE</span>)
  <span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">recountURLL: &quot;</span>, recountURL)
  
  ## Define the local file name
  localFile &lt;-<span class="st"> </span><span class="kw">file.path</span>(download.dir, <span class="kw">basename</span>(recountURL))
  <span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">localFile: &quot;</span>, localFile)

  ## Index local file for later use
  localFiles[type] &lt;-<span class="st"> </span>localFile

  
  ## Download the data only if required
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(localFile)) {
    <span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Dowloading &quot;</span>, type, <span class="st">&quot; from ReCount for study &quot;</span>, recountID)
    url &lt;-<span class="st"> </span><span class="kw">download_study</span>(
      recountID, 
      <span class="dt">outdir =</span> download.dir, 
      <span class="dt">type =</span> type)
  } <span class="cf">else</span> {
    <span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">file already there: &quot;</span>, localFile)
    
  }
}
## Download the gene-level RangedSummarizedExperiment data
<span class="co"># download_study(project = &quot;TCGA&quot;, type = &quot;counts-gene&quot;, outdir = download.dir)</span></code></pre></div>
<p>We downloaded the following file types from Recount2: rse-gene.</p>
</div>
<div id="data-loading" class="section level2">
<h2>Data loading</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># ## Choose a way to load the data. </span>
<span class="co"># ## either from the Rdata file or from the TSV files . </span>
<span class="co"># loadRdata &lt;- TRUE </span>
<span class="co"># </span>
<span class="co"># </span>
<span class="co"># if (loadRdata) {</span>
## Load the RData memory image provided by Recount2, whcih contains the count table + pheno table
<span class="kw">message</span>(<span class="st">&quot;Loading TCGA data from Recount&quot;</span>)
<span class="kw">load</span>(localFiles[<span class="st">&quot;rse-gene&quot;</span>])


## Extract the pheno table from the rse-gene object
phenoTable &lt;-<span class="st"> </span><span class="kw">colData</span>(rse_gene) ## phenotype per run
project.name &lt;-<span class="st"> &quot;Breast invasive carcinoma&quot;</span>
<span class="kw">message</span>(<span class="st">&quot;Identifying samples for project &quot;</span>, project.name)
project.samples &lt;-<span class="st"> </span>phenoTable<span class="op">$</span>gdc_cases.tissue_source_site.project <span class="op">==</span><span class="st"> </span>project.name
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Number of project samples: &quot;</span>, <span class="kw">sum</span>(project.samples))

## Select the columns relevant for the Breast cancer study
immuno.columns &lt;-<span class="st"> </span><span class="kw">c</span>(
  <span class="dt">ER =</span> <span class="st">&quot;xml_breast_carcinoma_estrogen_receptor_status&quot;</span>,
  <span class="dt">PR =</span> <span class="st">&quot;xml_breast_carcinoma_progesterone_receptor_status&quot;</span>,
  <span class="dt">HER2 =</span> <span class="st">&quot;xml_lab_proc_her2_neu_immunohistochemistry_receptor_status&quot;</span>)
<span class="kw">message</span>(<span class="st">&quot;Immuno marker columns:</span><span class="ch">\n\t</span><span class="st">&quot;</span>, <span class="kw">paste</span>(<span class="dt">collapse =</span> <span class="st">&quot;</span><span class="ch">\n\t</span><span class="st">&quot;</span>, immuno.columns))

## Only keep project samples with valid markers (i.e. Negative or Positive), discard Undertermined, Equivocal and  NA
valid.markers &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Negative&quot;</span>, <span class="st">&quot;Positive&quot;</span>)
<span class="kw">message</span>(<span class="st">&quot;Selecting samples with valid immuno markers: &quot;</span>, <span class="kw">paste</span>(<span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span>, valid.markers))
valid.samples &lt;-<span class="st"> </span>
<span class="st">  </span>project.samples <span class="op">&amp;</span>
<span class="st">  </span><span class="kw">apply</span>(<span class="kw">is.na</span>(phenoTable[, immuno.columns]), <span class="dv">1</span>, sum) <span class="op">==</span><span class="st"> </span><span class="dv">0</span> <span class="op">&amp;</span>
<span class="st">  </span>phenoTable[, immuno.columns[<span class="st">&quot;ER&quot;</span>]] <span class="op">%in%</span><span class="st"> </span>valid.markers <span class="op">&amp;</span>
<span class="st">  </span>phenoTable[, immuno.columns[<span class="st">&quot;PR&quot;</span>]] <span class="op">%in%</span><span class="st"> </span>valid.markers <span class="op">&amp;</span>
<span class="st">  </span>phenoTable[, immuno.columns[<span class="st">&quot;HER2&quot;</span>]] <span class="op">%in%</span><span class="st"> </span>valid.markers
<span class="co"># table(valid.samples)</span>
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">valid samples: &quot;</span>, <span class="kw">sum</span>(valid.samples))

## Subset the expression data by selecting the valid samples
<span class="kw">message</span>(<span class="st">&quot;Extracting expression data for valid samples&quot;</span>)
rse.valid.samples &lt;-<span class="st"> </span><span class="kw">subset</span>(rse_gene, <span class="dt">select =</span> valid.samples)  
<span class="co"># dim(rse.valid.samples)</span>

## Extract a pheno table for the subset
pheno &lt;-<span class="st"> </span><span class="kw">colData</span>(rse.valid.samples)
<span class="co"># dim(pheno.valid.samples)</span>

## Scale counts by mapped reads, in order to to get read counts per gene
<span class="co">#rse.valid.samples.scaled &lt;- scale_counts(rse.valid.samples, by = &quot;mapped_reads&quot;)</span>
<span class="co"># summary(assay(rse.valid.samples.scaled)[,1:7])</span>
## <span class="al">WARNING</span>: THIS DOES NOT WORK WITH THIS DATASET: ALL VALUES ARE SET TO 0.
## I THIS COMMENT IT

## Extract the count table from the rse-gene object
<span class="kw">message</span>(<span class="st">&quot;Extracting counts per gene&quot;</span>)
counts.per.gene &lt;-<span class="st"> </span><span class="kw">assay</span>(rse.valid.samples)
<span class="co"># summary(counts.per.gene[, 1:10])</span>
<span class="co"># dim(counts.per.gene)</span>


<span class="co"># dim(phenoTable)</span>
## <span class="al">NOTE</span>: this phenoTable is a complex object: a DataFrame whose cells may be a list containing a table. 
## To avoid messing around, I prefer to use the TSV version of the pheno table.
<span class="co"># dim(pheno)</span>
<span class="co"># class(pheno)</span>

<span class="co">#   </span>
<span class="co">#   # dim(countTable)</span>
<span class="co"># } else {</span>
<span class="co">#   ## Load the data from the TSV files downloaded from Recount2</span>
<span class="co">#   </span>
<span class="co">#   </span>
<span class="co">#   ## Load the TCGA count table.</span>
<span class="co">#   ## We use gzfile to directly load the zgipped file, </span>
<span class="co">#   ## and data.table:::fread() to accelerate the reading. </span>
<span class="co">#   ## system.time(counts.per.gene &lt;- read.delim(file = gzfile(localFiles[&quot;counts-gene&quot;]))) ## This takes 430 seconds</span>
<span class="co">#   message(&quot;Loading counts per gene&quot;)</span>
<span class="co">#   system.time(counts.per.gene &lt;- fread(file = localFiles[&quot;counts-gene&quot;]))</span>
<span class="co">#   # This takes 29 seconds</span>
<span class="co"># </span>
<span class="co">#   ## Load the TCGA pheno table</span>
<span class="co">#   message(&quot;Reading phenotype table&quot;)</span>
<span class="co">#   pheno &lt;- read.delim(file = localFiles[&quot;phenotype&quot;])</span>
<span class="co">#   # dim(pheno)</span>
<span class="co">#   </span>
<span class="co">#   # dim(counts.per.gene) </span>
<span class="co"># }</span>

  
<span class="kw">message</span>(<span class="st">&quot;Selected dataset (project and valid markers)&quot;</span>)
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Read count table contains &quot;</span>, <span class="kw">nrow</span>(counts.per.gene), <span class="st">&quot; rows (genes) x &quot;</span>, <span class="kw">ncol</span>(counts.per.gene), <span class="st">&quot; columns (samples). &quot;</span>)
<span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Pheno table contains &quot;</span>, <span class="kw">nrow</span>(pheno), <span class="st">&quot; rows (samples) x &quot;</span>, <span class="kw">ncol</span>(pheno), <span class="st">&quot; columns (sample attributes). &quot;</span>)</code></pre></div>
<p>The full TCGA gene count table contains 819 samples (columns) and 58037 genes (rows).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">sort</span>(<span class="kw">table</span>(phenoTable<span class="op">$</span>gdc_cases.project.name), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>),
      <span class="dt">caption =</span> <span class="st">&quot;Number of samples per TGCA project. &quot;</span>, 
      <span class="dt">col.names =</span> <span class="kw">c</span>(<span class="st">&quot;Project&quot;</span>,<span class="st">&quot;samples&quot;</span>))</code></pre></div>
<table>
<caption>
Number of samples per TGCA project.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Project
</th>
<th style="text-align:right;">
samples
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Breast Invasive Carcinoma
</td>
<td style="text-align:right;">
1246
</td>
</tr>
<tr>
<td style="text-align:left;">
Kidney Renal Clear Cell Carcinoma
</td>
<td style="text-align:right;">
616
</td>
</tr>
<tr>
<td style="text-align:left;">
Lung Adenocarcinoma
</td>
<td style="text-align:right;">
601
</td>
</tr>
<tr>
<td style="text-align:left;">
Uterine Corpus Endometrial Carcinoma
</td>
<td style="text-align:right;">
589
</td>
</tr>
<tr>
<td style="text-align:left;">
Thyroid Carcinoma
</td>
<td style="text-align:right;">
572
</td>
</tr>
<tr>
<td style="text-align:left;">
Prostate Adenocarcinoma
</td>
<td style="text-align:right;">
558
</td>
</tr>
<tr>
<td style="text-align:left;">
Lung Squamous Cell Carcinoma
</td>
<td style="text-align:right;">
555
</td>
</tr>
<tr>
<td style="text-align:left;">
Head and Neck Squamous Cell Carcinoma
</td>
<td style="text-align:right;">
548
</td>
</tr>
<tr>
<td style="text-align:left;">
Colon Adenocarcinoma
</td>
<td style="text-align:right;">
546
</td>
</tr>
<tr>
<td style="text-align:left;">
Brain Lower Grade Glioma
</td>
<td style="text-align:right;">
532
</td>
</tr>
<tr>
<td style="text-align:left;">
Skin Cutaneous Melanoma
</td>
<td style="text-align:right;">
473
</td>
</tr>
<tr>
<td style="text-align:left;">
Stomach Adenocarcinoma
</td>
<td style="text-align:right;">
453
</td>
</tr>
<tr>
<td style="text-align:left;">
Bladder Urothelial Carcinoma
</td>
<td style="text-align:right;">
433
</td>
</tr>
<tr>
<td style="text-align:left;">
Ovarian Serous Cystadenocarcinoma
</td>
<td style="text-align:right;">
430
</td>
</tr>
<tr>
<td style="text-align:left;">
Liver Hepatocellular Carcinoma
</td>
<td style="text-align:right;">
424
</td>
</tr>
<tr>
<td style="text-align:left;">
Kidney Renal Papillary Cell Carcinoma
</td>
<td style="text-align:right;">
323
</td>
</tr>
<tr>
<td style="text-align:left;">
Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma
</td>
<td style="text-align:right;">
309
</td>
</tr>
<tr>
<td style="text-align:left;">
Sarcoma
</td>
<td style="text-align:right;">
265
</td>
</tr>
<tr>
<td style="text-align:left;">
Esophageal Carcinoma
</td>
<td style="text-align:right;">
198
</td>
</tr>
<tr>
<td style="text-align:left;">
Pheochromocytoma and Paraganglioma
</td>
<td style="text-align:right;">
187
</td>
</tr>
<tr>
<td style="text-align:left;">
Pancreatic Adenocarcinoma
</td>
<td style="text-align:right;">
183
</td>
</tr>
<tr>
<td style="text-align:left;">
Rectum Adenocarcinoma
</td>
<td style="text-align:right;">
177
</td>
</tr>
<tr>
<td style="text-align:left;">
Glioblastoma Multiforme
</td>
<td style="text-align:right;">
175
</td>
</tr>
<tr>
<td style="text-align:left;">
Testicular Germ Cell Tumors
</td>
<td style="text-align:right;">
156
</td>
</tr>
<tr>
<td style="text-align:left;">
Acute Myeloid Leukemia
</td>
<td style="text-align:right;">
126
</td>
</tr>
<tr>
<td style="text-align:left;">
Thymoma
</td>
<td style="text-align:right;">
122
</td>
</tr>
<tr>
<td style="text-align:left;">
Kidney Chromophobe
</td>
<td style="text-align:right;">
91
</td>
</tr>
<tr>
<td style="text-align:left;">
Mesothelioma
</td>
<td style="text-align:right;">
87
</td>
</tr>
<tr>
<td style="text-align:left;">
Uveal Melanoma
</td>
<td style="text-align:right;">
80
</td>
</tr>
<tr>
<td style="text-align:left;">
Adrenocortical Carcinoma
</td>
<td style="text-align:right;">
79
</td>
</tr>
<tr>
<td style="text-align:left;">
Uterine Carcinosarcoma
</td>
<td style="text-align:right;">
57
</td>
</tr>
<tr>
<td style="text-align:left;">
Lymphoid Neoplasm Diffuse Large B-cell Lymphoma
</td>
<td style="text-align:right;">
48
</td>
</tr>
<tr>
<td style="text-align:left;">
Cholangiocarcinoma
</td>
<td style="text-align:right;">
45
</td>
</tr>
</tbody>
</table>
<p>For the study case, we select the project <strong>1246</strong>, which contains 1246 samples.</p>
</div>
<div id="class-labels" class="section level2">
<h2>Class labels</h2>
<p>The class of cancer is defined by combining three immunological markers:</p>
<ul>
<li>HER2,</li>
<li>ER (estrogen receptor)</li>
<li>Pr (progesterone receptor)</li>
</ul>
<p>The tables below provide summaries of the marker status for all samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">project.markers &lt;-<span class="st"> </span>phenoTable[project.samples, immuno.columns]
<span class="kw">colnames</span>(project.markers) &lt;-<span class="st"> </span><span class="kw">names</span>(immuno.columns)
<span class="kw">summary</span>(<span class="kw">as.data.frame</span>(project.markers))</code></pre></div>
<pre><code>             ER                  PR                 HER2    
              :  0                :  0                :  0  
 Indeterminate:  3   Indeterminate:  5   Equivocal    :201  
 Negative     :273   Negative     :387   Indeterminate: 12  
 Positive     :910   Positive     :793   Negative     :636  
 NA&#39;s         : 60   NA&#39;s         : 61   Positive     :190  
                                         NA&#39;s         :207  </code></pre>
<p>Marker status summary after having discarded invalid values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">valid.markers &lt;-<span class="st"> </span>phenoTable[valid.samples, immuno.columns]
<span class="kw">colnames</span>(valid.markers) &lt;-<span class="st"> </span><span class="kw">names</span>(immuno.columns)
<span class="kw">summary</span>(<span class="kw">as.data.frame</span>(valid.markers))</code></pre></div>
<pre><code>             ER                  PR                 HER2    
              :  0                :  0                :  0  
 Indeterminate:  0   Indeterminate:  0   Equivocal    :  0  
 Negative     :184   Negative     :267   Indeterminate:  0  
 Positive     :635   Positive     :552   Negative     :631  
                                         Positive     :188  </code></pre>
<p>We discard the samples for which any of the markers is undefined (<code>NA</code>), “Indeterminate” or “Equivocal”. In total, this leaves us with 819 samples.</p>
<p>We then assign a class label to each sample, indicating its cancer subtype, which is assigned based on the combination of the three immunological markers.</p>
<div class="figure">
<img src="images/breast_cancer_typology.png" alt="Figure 1. Definition of breast cancer subtypes based on immunological markers HER2, ER (Estrogen Receptor), and PR (Progesterone Receptor)" />
<p class="caption">Figure 1. <strong>Definition of breast cancer subtypes</strong> based on immunological markers HER2, ER (Estrogen Receptor), and PR (Progesterone Receptor)</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Define sample classes based on the combination of 3 marker values
pheno<span class="op">$</span>cancer.type &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&quot;Unclassified&quot;</span>, <span class="dt">length.out =</span> <span class="kw">nrow</span>(pheno))

luminal.A &lt;-<span class="st"> </span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;ER&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Positive&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;PR&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Positive&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;HER2&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Negative&quot;</span>
pheno[luminal.A, <span class="st">&quot;cancer.type&quot;</span>] &lt;-<span class="st"> &quot;Luminal.A&quot;</span>

luminal.B &lt;-<span class="st"> </span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;ER&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Positive&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;PR&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Positive&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;HER2&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Positive&quot;</span>
pheno[luminal.B, <span class="st">&quot;cancer.type&quot;</span>] &lt;-<span class="st"> &quot;Luminal.B&quot;</span>

her2plus &lt;-<span class="st"> </span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;ER&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Negative&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;PR&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Negative&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;HER2&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Positive&quot;</span>
pheno[her2plus, <span class="st">&quot;cancer.type&quot;</span>] &lt;-<span class="st"> &quot;HER2pos&quot;</span>

basal.like &lt;-<span class="st"> </span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;ER&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Negative&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;PR&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Negative&quot;</span> <span class="op">&amp;</span>
<span class="st">  </span>pheno[, immuno.columns[<span class="st">&quot;HER2&quot;</span>]] <span class="op">==</span><span class="st"> &quot;Negative&quot;</span>
pheno[basal.like, <span class="st">&quot;cancer.type&quot;</span>] &lt;-<span class="st"> &quot;Basal.like&quot;</span>

<span class="kw">kable</span>(<span class="kw">sort</span>(<span class="dt">decreasing =</span> <span class="ot">TRUE</span>, <span class="kw">table</span>(pheno[, <span class="st">&quot;cancer.type&quot;</span>])), <span class="dt">useNA =</span> <span class="st">&quot;ifany&quot;</span>, <span class="dt">caption =</span> <span class="st">&quot;Number of samples per cancer subtype. Cancer subtypes were defined based on the combination of HER2, ER and PR markers. &quot;</span>)</code></pre></div>
<table>
<caption>
Number of samples per cancer subtype. Cancer subtypes were defined based on the combination of HER2, ER and PR markers.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Var1
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Luminal.A
</td>
<td style="text-align:right;">
422
</td>
</tr>
<tr>
<td style="text-align:left;">
Basal.like
</td>
<td style="text-align:right;">
131
</td>
</tr>
<tr>
<td style="text-align:left;">
Luminal.B
</td>
<td style="text-align:right;">
118
</td>
</tr>
<tr>
<td style="text-align:left;">
Unclassified
</td>
<td style="text-align:right;">
107
</td>
</tr>
<tr>
<td style="text-align:left;">
HER2pos
</td>
<td style="text-align:right;">
41
</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Get the identifiers of the selected samples
selected.sample.ids &lt;-<span class="st"> </span><span class="kw">rownames</span>(pheno)
<span class="co"># summary(selected.sample.ids %in% colnames(counts.per.gene))</span>
<span class="co"># summary(colnames(counts.per.gene) %in% selected.sample.ids)</span>

## Get the subset of the big count table corresponding to our selected IDs
## Note the special notation due to the specific class of the fread() result.
<span class="co"># if (!loadRdata) {</span>
<span class="co">#   selected.counts &lt;- as.data.frame(counts.per.gene[, ..selected.sample.ids])</span>
<span class="co"># }</span>
<span class="co"># dim(selected.counts)</span>
<span class="co"># summary(selected.counts[, 1:10])</span>
<span class="co"># summary(counts.per.gene[, 1:10])</span></code></pre></div>
<hr />
</div>
<div id="gene-filtering" class="section level2">
<h2>Gene filtering</h2>
<div id="percent-zero-counts" class="section level3">
<h3>Percent zero counts</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Discard genes having zeros in at least 95% of samples
<span class="kw">message</span>(<span class="st">&quot;Applying threshold on the percent of non-zero counts per gene: &quot;</span>, parameters<span class="op">$</span>gene.filter.percent.zeros, <span class="st">&quot;%&quot;</span>)
percent.zeros &lt;-<span class="st"> </span><span class="dv">100</span><span class="op">*</span><span class="kw">apply</span>(counts.per.gene <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="dv">1</span>, sum) <span class="op">/</span><span class="st"> </span><span class="kw">ncol</span>(counts.per.gene)
<span class="kw">hist</span>(percent.zeros, <span class="dt">breaks =</span> <span class="dv">20</span>, <span class="dt">col =</span> <span class="st">&quot;#DDDDDD&quot;</span>, 
     <span class="dt">main =</span> <span class="st">&quot;Filter on the percentage of zero counts&quot;</span>, 
     <span class="dt">xlab =</span> <span class="st">&quot;Percent of samples&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>)
<span class="kw">arrows</span>(<span class="dt">x0 =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
       <span class="dt">y0 =</span> <span class="dv">10000</span>, 
       <span class="dt">x1 =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
       <span class="dt">y1 =</span> <span class="dv">6000</span>, 
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>,
       <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">length =</span> <span class="fl">0.1</span>)
<span class="kw">text</span>(<span class="dt">x =</span> parameters<span class="op">$</span>gene.filter.percent.zeros, 
     <span class="dt">y =</span> <span class="dv">10000</span>, 
     <span class="dt">labels =</span> <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, parameters<span class="op">$</span>gene.filter.percent.zeros, <span class="st">&quot;%&quot;</span>),
     <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">pos =</span> <span class="dv">3</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TCGA_BIC_gene_filtering_zero_counts-1.png" alt="Frequency of samples with zero counts per gene. Genes exceeding the thresold (red arrow) were filtered out. " width="80%" />
<p class="caption">
Frequency of samples with zero counts per gene. Genes exceeding the thresold (red arrow) were filtered out.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">min.count &lt;-<span class="st"> </span><span class="kw">apply</span>(counts.per.gene, <span class="dv">1</span>, min)
<span class="kw">summary</span>(min.count)</code></pre></div>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
      0       0       0    6824     938 2168821 </code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(<span class="kw">log2</span>(min.count <span class="op">+</span><span class="st"> </span>parameters<span class="op">$</span>epsilon), <span class="dt">breaks =</span> <span class="dv">100</span>, <span class="dt">col =</span> <span class="st">&quot;#BBBBFF&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;log2(min count per gene)&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>, <span class="dt">las =</span> <span class="dv">1</span>,
     <span class="dt">main =</span> <span class="st">&quot;Filtering on min count per gene&quot;</span>)
<span class="kw">arrows</span>(<span class="dt">x0 =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
       <span class="dt">y0 =</span> <span class="dv">3000</span>, 
       <span class="dt">x1 =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
       <span class="dt">y1 =</span> <span class="dv">1000</span>, 
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">lwd =</span> <span class="dv">2</span>,
       <span class="dt">angle =</span> <span class="dv">30</span>, <span class="dt">length =</span> <span class="fl">0.1</span>)
<span class="kw">text</span>(<span class="dt">x =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count), 
     <span class="dt">y =</span> <span class="dv">3000</span>, 
     <span class="dt">labels =</span> <span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, <span class="st">&quot;log2(&quot;</span>, 
                    parameters<span class="op">$</span>gene.filter.min.count,
                    <span class="st">&quot;) = &quot;</span>, <span class="kw">signif</span>(<span class="dt">digits =</span> <span class="dv">3</span>, <span class="kw">log2</span>(parameters<span class="op">$</span>gene.filter.min.count))),
     <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">pos =</span> <span class="dv">3</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TCGA_BIC_gene_filtering_min_count-1.png" alt="Distribution of min counts per gene. Genes below the thresold (red arrow) are filtered out. " width="80%" />
<p class="caption">
Distribution of min counts per gene. Genes below the thresold (red arrow) are filtered out.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filtered.out.genes &lt;-<span class="st"> </span>
<span class="st">  </span>percent.zeros <span class="op">&gt;</span><span class="st"> </span>parameters<span class="op">$</span>gene.filter.percent.zeros <span class="op">|</span>
<span class="st">  </span>min.count <span class="op">&lt;</span><span class="st"> </span>parameters<span class="op">$</span>gene.filter.min.count
<span class="co"># table(kept.genes)</span>

## Genes passing the filters
filtered.genes &lt;-<span class="st"> </span><span class="op">!</span>filtered.out.genes
filtered.counts &lt;-<span class="st"> </span>counts.per.gene[filtered.genes, ]</code></pre></div>
<p>We discard “undetected” genes, i.e. those having zero counts in at least 95 percent of the samples, or those with a maximal count inferior to 10. This led us to discard 37788 genes.</p>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2>Differential expression</h2>
<p>We select differentially expressed genes in order to produce a file with a restricted number of genes for clustering.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Use the DESeqDataSetFromMatrix to create a DESeqDataSet object
<span class="kw">message</span>(<span class="st">&quot;Creating DESeq2 dataset&quot;</span>)
dds0 &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(
  <span class="dt">countData =</span> <span class="kw">as.data.frame</span>(filtered.counts), 
  <span class="dt">colData =</span> <span class="kw">data.frame</span>(<span class="dt">cancer.type =</span> <span class="kw">as.factor</span>(pheno[, <span class="st">&quot;cancer.type&quot;</span>])), 
  <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>cancer.type)

<span class="co"># print(dds0) # Have a look at the short description of the DESeqDataSet</span>

## PROBLEM: the differential analysis with DESeq2 is stuck. 
<span class="co">#@ Probably because there are too many samples ?</span>
## TO DO: 
## - test it on IFB cluster
## - test edgeR as selection method
## - select a random subset of the samples

<span class="co"># ## Run  differential expression analysis with DESeq2</span>
<span class="co"># message(&quot;Running differential expression analysis with DESeq2&quot;)</span>
<span class="co"># dds0 &lt;- DESeq(dds0)</span>
<span class="co"># </span>
<span class="co"># # Cast the results from DESeq differential expression analysis</span>
<span class="co"># DEGtable &lt;- results(dds0, independentFiltering = FALSE)</span></code></pre></div>
</div>
<div id="normalization-of-the-counts-per-gene" class="section level2">
<h2>Normalization of the counts per gene</h2>
<p>We use DESeq2 function <code>estimateSizeFactors()</code> to estimate the library-wise size factors, which will be used to standardize the counts.</p>
<p>We then apply a log2 transformation in order to normalize these standardized counts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Normalizing using the method for an object of class&quot;CountDataSet&quot; 
dds.norm &lt;-<span class="st">  </span><span class="kw">estimateSizeFactors</span>(dds0)
<span class="co"># sizeFactors(dds.norm)</span>

counts.norm &lt;-<span class="st"> </span><span class="kw">counts</span>(dds.norm) <span class="op">+</span><span class="st"> </span>parameters<span class="op">$</span>epsilon
<span class="co"># dim(counts.norm)</span>
counts.log2.norm &lt;-<span class="st"> </span><span class="kw">log2</span>(counts.norm)
<span class="co"># dim(counts.log2.norm)</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">log2.count.breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(
  <span class="dt">from =</span> <span class="kw">log2</span>(parameters<span class="op">$</span>epsilon), 
  <span class="dt">to =</span> <span class="dv">32</span>, 
  <span class="dt">by =</span> <span class="fl">0.4</span>)

<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))

<span class="kw">hist</span>(<span class="kw">unlist</span>(<span class="kw">log2</span>(<span class="kw">unlist</span>(counts.per.gene <span class="op">+</span><span class="st"> </span>parameters<span class="op">$</span>epsilon))), <span class="dt">main =</span> <span class="st">&quot;Distribution of raw counts&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;log2(raw counts + epsilon)&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;Number of measures&quot;</span>,
     <span class="dt">breaks =</span> log2.count.breaks, 
     <span class="dt">col =</span> <span class="st">&quot;#FFDDBB&quot;</span>)

<span class="kw">hist</span>(<span class="kw">unlist</span>(counts.log2.norm), <span class="dt">main =</span> <span class="st">&quot;Distribution of normalized counts&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;log2(standardized counts)&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;Number of measures&quot;</span>,
     <span class="dt">breaks =</span> log2.count.breaks, 
     <span class="dt">col =</span> <span class="st">&quot;#DDFFBB&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/TCGA_BIC_norm_count_hist-1.png" alt="**Histogram of all counts.** (a) log2(raw counts + epsilon). The epsilon is added to avoid -Inf values for log2-transformed zeros. (b) Normalised counts. Normalization consists in scaling counts in order to ensure library size standardization, followed by a  log2 transformation. " width="80%" />
<p class="caption">
<strong>Histogram of all counts.</strong> (a) log2(raw counts + epsilon). The epsilon is added to avoid -Inf values for log2-transformed zeros. (b) Normalised counts. Normalization consists in scaling counts in order to ensure library size standardization, followed by a log2 transformation.
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</code></pre></div>
<hr />
</div>
<div id="exported-tables" class="section level2">
<h2>Exported tables</h2>
<p>The selected counts per gene and pheno table were exported in tab-separated value (TSV) format.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno.export.columns &lt;-<span class="st"> </span><span class="kw">names</span>(pheno)
<span class="cf">for</span> (col <span class="cf">in</span> pheno.export.columns) {
  ## Discard columns containing only NA values
  nona &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="op">!</span><span class="kw">is.na</span>(pheno[, col]))
  <span class="cf">if</span> (nona <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) {
    <span class="co"># message(&quot;\tDiscarding NA-only pheno column\t&quot;, col, &quot;\tnoNA = &quot;, nona)</span>
    pheno.export.columns &lt;-<span class="st"> </span><span class="kw">setdiff</span>(pheno.export.columns, col)
  }
  
  <span class="co"># Discard columns containing muli-value lists</span>
  <span class="cf">if</span> (<span class="kw">class</span>(pheno[, col]) <span class="op">==</span><span class="st"> &quot;list&quot;</span>) {
    <span class="kw">message</span>(<span class="st">&quot;</span><span class="ch">\t</span><span class="st">Discarding list-type pheno column</span><span class="ch">\t</span><span class="st">&quot;</span>, col)
    pheno.export.columns &lt;-<span class="st"> </span><span class="kw">setdiff</span>(pheno.export.columns, col)
    
  }
}
<span class="kw">message</span>(<span class="st">&quot;Exporting &quot;</span>, <span class="kw">length</span>(pheno.export.columns), <span class="st">&quot; pheno columns among &quot;</span>, <span class="kw">ncol</span>(pheno))

## Export the non-problematic columns of the pheno table
pheno.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, project.acronym, <span class="st">&quot;_pheno.tsv.gz&quot;</span>)
pheno.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(export.dir, pheno.file)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  pheno.path &lt;-<span class="st"> </span><span class="kw">gzfile</span>(pheno.path, <span class="st">&quot;w&quot;</span>) ## Compress file
}
<span class="kw">write.table</span>(<span class="dt">x =</span> <span class="kw">as.data.frame</span>(pheno[, pheno.export.columns]), 
            <span class="dt">file =</span> pheno.path, 
            <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  <span class="kw">close</span>(pheno.path)
}

## Export a small subset of relevant fields from the pheno table
pheno.selected.columns &lt;-<span class="st"> </span><span class="kw">c</span>(
  <span class="st">&quot;cancer.type&quot;</span>,
  immuno.columns
)
sample.class.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, project.acronym, <span class="st">&quot;_sample-classes.tsv.gz&quot;</span>)
sample.class.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(export.dir, sample.class.file)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  sample.class.path &lt;-<span class="st"> </span><span class="kw">gzfile</span>(sample.class.path, <span class="st">&quot;w&quot;</span>) ## Compress file
}
<span class="kw">write.table</span>(<span class="dt">x =</span> <span class="kw">as.data.frame</span>(pheno[, pheno.selected.columns]), 
            <span class="dt">file =</span> sample.class.path, 
            <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  <span class="kw">close</span>(sample.class.path)
}


## Export counts per gene
count.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, project.acronym, <span class="st">&quot;_counts_all-genes.tsv&quot;</span>)
count.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(export.dir, count.file)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  count.path &lt;-<span class="st"> </span><span class="kw">gzfile</span>(count.path, <span class="st">&quot;w&quot;</span>) ## Compress file
}
<span class="kw">write.table</span>(<span class="dt">x =</span> counts.per.gene, 
            <span class="dt">file =</span> count.path,
            <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  <span class="kw">close</span>(count.path)
}

## Export raw counts
filtered.count.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, project.acronym, <span class="st">&quot;_counts_filtered-genes.tsv&quot;</span>)
filtered.count.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(export.dir, filtered.count.file)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  filtered.count.path &lt;-<span class="st"> </span><span class="kw">gzfile</span>(filtered.count.path, <span class="st">&quot;w&quot;</span>) ## Compress file
}
<span class="kw">write.table</span>(<span class="dt">x =</span> filtered.counts, 
            <span class="dt">file =</span> filtered.count.path,
            <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  <span class="kw">close</span>(filtered.count.path)
}

<span class="co"># list.files(export.dir)</span>

## Export normalized counts
norm.count.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, project.acronym, <span class="st">&quot;_log2-norm-counts_all-genes.tsv&quot;</span>)
norm.count.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(export.dir, norm.count.file)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  norm.count.path &lt;-<span class="st"> </span><span class="kw">gzfile</span>(norm.count.path, <span class="st">&quot;w&quot;</span>) ## Compress file
}
<span class="kw">write.table</span>(<span class="dt">x =</span> counts.log2.norm, 
            <span class="dt">file =</span> norm.count.path,
            <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  <span class="kw">close</span>(norm.count.path)
}

## Export normalized counts
filtered.norm.count.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="dt">sep =</span> <span class="st">&quot;&quot;</span>, project.acronym, <span class="st">&quot;_log2-norm-counts_filtered-genes.tsv&quot;</span>)
filtered.norm.count.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(export.dir, filtered.norm.count.file)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  filtered.norm.count.path &lt;-<span class="st"> </span><span class="kw">gzfile</span>(filtered.norm.count.path, <span class="st">&quot;w&quot;</span>) ## Compress file
}
<span class="co"># dim(counts.log2.norm)</span>
<span class="kw">write.table</span>(<span class="dt">x =</span> counts.log2.norm, 
            <span class="dt">file =</span> filtered.norm.count.path,
            <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">row.names =</span> <span class="ot">TRUE</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)
<span class="cf">if</span> (parameters<span class="op">$</span>gzip.output.files) {
  <span class="kw">close</span>(filtered.norm.count.path)
}</code></pre></div>
<table>
<colgroup>
<col width="56%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th>Contants</th>
<th>File name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Export directory</td>
<td><a href="~/TCGA_import/data/BIC" class="uri">~/TCGA_import/data/BIC</a></td>
</tr>
<tr class="even">
<td>Pheno table</td>
<td><a href="~/TCGA_import/data/BIC" class="uri">~/TCGA_import/data/BIC</a></td>
</tr>
<tr class="odd">
<td>Counts per gene (all genes)</td>
<td><a href="BIC_counts_all-genes.tsv" class="uri">BIC_counts_all-genes.tsv</a></td>
</tr>
<tr class="even">
<td>Counts per gene (filtered genes)</td>
<td><a href="BIC_counts_filtered-genes.tsv" class="uri">BIC_counts_filtered-genes.tsv</a></td>
</tr>
<tr class="odd">
<td>Normalized counts per gene (all genes)</td>
<td><a href="BIC_log2-norm-counts_all-genes.tsv" class="uri">BIC_log2-norm-counts_all-genes.tsv</a></td>
</tr>
<tr class="even">
<td>Normalized counts per gene (filtered genes)</td>
<td><a href="BIC_log2-norm-counts_filtered-genes.tsv" class="uri">BIC_log2-norm-counts_filtered-genes.tsv</a></td>
</tr>
</tbody>
</table>
<hr />
<p>Contact: <a href="mailto:Jacques.van-Helden@univ-amu.fr">Jacques.van-Helden@univ-amu.fr</a></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
