<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Olivier Sand and Jacques van Helden" />

<meta name="date" content="2020-05-27" />

<title>DU-Bii Study cases - Mouse fibrotic kidney</title>

<script src="prepare-data_pavkovic_2019_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="prepare-data_pavkovic_2019_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="prepare-data_pavkovic_2019_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="prepare-data_pavkovic_2019_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="prepare-data_pavkovic_2019_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="prepare-data_pavkovic_2019_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="prepare-data_pavkovic_2019_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="prepare-data_pavkovic_2019_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="prepare-data_pavkovic_2019_files/navigation-1.1/tabsets.js"></script>
<script src="prepare-data_pavkovic_2019_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DU-Bii Study cases - Mouse fibrotic kidney</h1>
<h3 class="subtitle">DUBii 2019</h3>
<h4 class="author">Olivier Sand and Jacques van Helden</h4>
<h4 class="date">2020-05-27</h4>

</div>


<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">## Define parameters for the analysis</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">outdirs &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">outdirs<span class="op">$</span>main &lt;-<span class="st"> </span><span class="kw">getwd</span>()</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">## Output directory</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6">outdirs<span class="op">$</span>data &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdirs<span class="op">$</span>main, <span class="st">&quot;Mus_musculus&quot;</span> , <span class="st">&quot;pavkovic_2019&quot;</span>, <span class="st">&quot;data&quot;</span>)</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="kw">dir.create</span>(outdirs<span class="op">$</span>data, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">## Analysis parameters</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10">epsilon &lt;-<span class="st"> </span><span class="fl">0.1</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co">## Original graphical parameters</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13">par.ori &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="dt">no.readonly =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<div id="reference" class="section level2">
<h2>Reference</h2>
<p>Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019).</p>
<ul>
<li><a href="https://doi.org/10.1038/s41597-019-0095-5" class="uri">https://doi.org/10.1038/s41597-019-0095-5</a></li>
<li><a href="https://www.nature.com/articles/s41597-019-0095-5#citeas" class="uri">https://www.nature.com/articles/s41597-019-0095-5#citeas</a></li>
<li>Mouse fibrotic kidney browser: <a href="http://hbcreports.med.harvard.edu/fmm/" class="uri">http://hbcreports.med.harvard.edu/fmm/</a></li>
<li>Data on Zenodo: <a href="https://zenodo.org/record/2592516" class="uri">https://zenodo.org/record/2592516</a></li>
</ul>
</div>
<div id="data-sources" class="section level2">
<h2>Data sources</h2>
<table>
<colgroup>
<col width="52%" />
<col width="47%" />
</colgroup>
<thead>
<tr class="header">
<th>Doc</th>
<th>URL</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Total RNA for the experiment on Unilateral ureter obstruction (UUO) model</td>
<td></td>
</tr>
<tr class="even">
<td><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118339" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118339</a></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="download-transcriptome-data" class="section level2">
<h2>Download transcriptome data</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#series &lt;- &quot;GSE118341&quot;</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co">#transcriptomeSource &lt;- paste0(&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=&quot;, series, &quot;&amp;format=file&quot;)</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">## BEWARE, the extension is wrong in Zenodo: this is a tar file, not a zip !</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5">archiveFile &lt;-<span class="st"> &quot;MouseKidneyFibrOmics-v1.0.zip&quot;</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">archiveURL &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&quot;https://zenodo.org/record/2592516/files/hbc&quot;</span>, archiveFile)</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">localDataArchive &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdirs<span class="op">$</span>data, archiveFile)</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="cf">if</span> (<span class="kw">file.exists</span>(localDataArchive)) {</a>
<a class="sourceLine" id="cb2-10" data-line-number="10">  <span class="kw">message</span>(<span class="st">&quot;Data archive already downloaded: &quot;</span>, localDataArchive)</a>
<a class="sourceLine" id="cb2-11" data-line-number="11">} <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">  <span class="kw">message</span>(<span class="st">&quot;Downloading data archive from zenodo: &quot;</span>, archiveURL)</a>
<a class="sourceLine" id="cb2-13" data-line-number="13">  <span class="kw">download.file</span>(<span class="dt">url =</span> archiveURL, <span class="dt">destfile =</span>  localDataArchive)</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">}</a>
<a class="sourceLine" id="cb2-15" data-line-number="15"></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">## Define destination directory</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co"># outdirs$csv &lt;- file.path(outdirs$data, &quot;CSV&quot;)</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co"># dir.create(outdirs$csv, showWarnings = FALSE, recursive = TRUE)</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">## Uncompess the archive</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="kw">unzip</span>(<span class="dt">zipfile =</span> localDataArchive, <span class="dt">exdir =</span> outdirs<span class="op">$</span>data)</a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="kw">setwd</span>(outdirs<span class="op">$</span>main)</a></code></pre></div>
</div>
<div id="load-raw-counts" class="section level2">
<h2>Load raw counts</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#### Load raw counts data table ####</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">rawCountFile &lt;-<span class="st"> </span><span class="kw">file.path</span>(</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">  outdirs<span class="op">$</span>data,</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">  <span class="st">&quot;hbc-MouseKidneyFibrOmics-a39e55a/tables/uuo/results/counts/raw_counts.csv.gz&quot;</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">rawCounts &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> rawCountFile, <span class="dt">header =</span> <span class="dv">1</span>, <span class="dt">row.names =</span> <span class="dv">1</span>)</a></code></pre></div>
<p>The RNA-seq trnascriptome data was loaded as raw counts. This table contains 46679 rows (genes) and 15 columns (samples).</p>
</div>
<div id="distribution-of-raw-counts" class="section level2">
<h2>Distribution of raw counts</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">#### Count distrib ####</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># head(rawCounts)</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw">hist</span>(<span class="kw">unlist</span>(rawCounts), <span class="dt">breaks =</span> <span class="dv">1000</span>, </a>
<a class="sourceLine" id="cb4-6" data-line-number="6">     <span class="dt">main =</span> <span class="st">&quot;Raw count distribution&quot;</span>, </a>
<a class="sourceLine" id="cb4-7" data-line-number="7">     <span class="dt">xlab =</span> <span class="st">&quot;Raw counts&quot;</span>, </a>
<a class="sourceLine" id="cb4-8" data-line-number="8">     <span class="dt">ylab =</span> <span class="st">&quot;Number of genes (all samples)&quot;</span>)</a>
<a class="sourceLine" id="cb4-9" data-line-number="9"></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="kw">hist</span>(<span class="kw">unlist</span>(rawCounts), <span class="dt">breaks =</span> <span class="dv">500000</span>, </a>
<a class="sourceLine" id="cb4-11" data-line-number="11">     <span class="dt">main =</span> <span class="st">&quot;Raw count distribution&quot;</span>, </a>
<a class="sourceLine" id="cb4-12" data-line-number="12">     <span class="dt">xlab =</span> <span class="st">&quot;Raw counts (truncated abscissa&quot;</span>, </a>
<a class="sourceLine" id="cb4-13" data-line-number="13">     <span class="dt">ylab =</span> <span class="st">&quot;Number of genes (all samples)&quot;</span>,</a>
<a class="sourceLine" id="cb4-14" data-line-number="14">     <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">500</span>))</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="figures/mouse-kidney_RNA-seq_count_distrib-1.png" alt="Distribution of raw counts" width="100%" />
<p class="caption">
Distribution of raw counts
</p>
</div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</a></code></pre></div>
<p>The distribution of raw counts is not very informative, because the range is defined by some outlier, i.e. a gene having a huge number of reads. Even with strong zoom on the abcsissa range from 0 to 500, the histogram shows a steep drop in the first bins.</p>
</div>
<div id="log2-transformed-counts" class="section level2">
<h2>Log2-transformed counts</h2>
<p>A typical approach is to normalise the counts by applying a log2 transformation . This however creates a problem when the counts of a given gene in a given sample is 0. To circumvent this, we can add an epsilon before the log2 transformation.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co">#### Log2 transformatiojn of the counts ####</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">log2Counts &lt;-<span class="st"> </span><span class="kw">log2</span>(rawCounts <span class="op">+</span><span class="st"> </span>epsilon)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="kw">hist</span>(<span class="kw">unlist</span>(log2Counts), <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">-5</span>, <span class="dt">to =</span> <span class="dv">22</span>, <span class="dt">by =</span> <span class="fl">0.1</span>),</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">     <span class="dt">main =</span> <span class="st">&quot;log2-counts distribution&quot;</span>,</a>
<a class="sourceLine" id="cb6-7" data-line-number="7">     <span class="dt">xlab =</span> <span class="st">&quot;log2(Counts + epsilon)&quot;</span>, </a>
<a class="sourceLine" id="cb6-8" data-line-number="8">     <span class="dt">ylab =</span> <span class="st">&quot;Number of genes&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;#BBFFBB&quot;</span>)</a></code></pre></div>
<p><img src="figures/mouse-kidney_log2_counts-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="box-plots" class="section level2">
<h2>Box plots</h2>
<p>We can now inspect the distribution of counts per sample with the <code>boxplot()</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">#### Box plots ####</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">5</span>, <span class="dv">1</span>))</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="kw">boxplot</span>(log2Counts, <span class="dt">horizontal =</span> <span class="ot">TRUE</span>, <span class="dt">las =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="figures/mouse-kidney_box%20plots-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We notice an obvious problem: the vast majority of counts is very small. This can result from different causes, whcih will not be investigted in this contet.</p>
</div>
<div id="gene-filtering" class="section level2">
<h2>Gene filtering</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">undetectedGenes &lt;-<span class="st"> </span><span class="kw">apply</span>(rawCounts, <span class="dt">MARGIN =</span> <span class="dv">1</span>, <span class="dt">FUN =</span> sum) <span class="op">==</span><span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw">table</span>(undetectedGenes)</a></code></pre></div>
<pre><code>undetectedGenes
FALSE  TRUE 
33379 13300 </code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">log2CountsFiltered &lt;-<span class="st"> </span>log2Counts[<span class="op">!</span>undetectedGenes, ]</a></code></pre></div>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>R version 4.0.0 (2020-04-24)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.28

loaded via a namespace (and not attached):
 [1] compiler_4.0.0  magrittr_1.5    tools_4.0.0     htmltools_0.4.0 yaml_2.2.1      Rcpp_1.0.4.6    stringi_1.4.6   rmarkdown_2.1   highr_0.8       stringr_1.4.0   xfun_0.14       digest_0.6.25   rlang_0.4.6     evaluate_0.14  </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
