<!DOCTYPE html>
<html>
<head>
  <title>DU-Bii Study cases - Mouse fibrotic kidney</title>

  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <meta name="generator" content="pandoc" />



  <meta name="date" content="2020-05-27" />

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">

  <base target="_blank">

  <script type="text/javascript">
    var SLIDE_CONFIG = {
      // Slide settings
      settings: {
                title: 'DU-Bii Study cases - Mouse fibrotic kidney',
                        subtitle: 'DUBii 2020',
                useBuilds: true,
        usePrettify: true,
        enableSlideAreas: true,
        enableTouch: true,
                      },

      // Author information
      presenters: [
            {
        name:  'Olivier Sand and Jacques van Helden' ,
        company: '',
        gplus: '',
        twitter: '',
        www: '',
        github: ''
      },
            ]
    };
  </script>

  <link href="prepare-data_pavkovic_2019_files/ioslides-13.5.1/fonts/fonts.css" rel="stylesheet" />
  <link href="prepare-data_pavkovic_2019_files/ioslides-13.5.1/theme/css/default.css" rel="stylesheet" />
  <link href="prepare-data_pavkovic_2019_files/ioslides-13.5.1/theme/css/phone.css" rel="stylesheet" />
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/modernizr.custom.45394.js"></script>
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/prettify/prettify.js"></script>
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/prettify/lang-r.js"></script>
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/prettify/lang-yaml.js"></script>
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/hammer.js"></script>
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/slide-controller.js"></script>
  <script src="prepare-data_pavkovic_2019_files/ioslides-13.5.1/js/slide-deck.js"></script>

  <style type="text/css">

    b, strong {
      font-weight: bold;
    }

    em {
      font-style: italic;
    }

    summary {
      display: list-item;
    }

    slides > slide {
      -webkit-transition: all 0.4s ease-in-out;
      -moz-transition: all 0.4s ease-in-out;
      -o-transition: all 0.4s ease-in-out;
      transition: all 0.4s ease-in-out;
    }

    .auto-fadein {
      -webkit-transition: opacity 0.6s ease-in;
      -webkit-transition-delay: 0.4s;
      -moz-transition: opacity 0.6s ease-in 0.4s;
      -o-transition: opacity 0.6s ease-in 0.4s;
      transition: opacity 0.6s ease-in 0.4s;
      opacity: 0;
    }
/* https://github.com/ropensci/plotly/pull/524#issuecomment-468142578 */
slide:not(.current) .plotly.html-widget{
  display: block;
}

  </style>


</head>

<body style="opacity: 0">

<slides class="layout-widescreen">

  <slide class="title-slide segue nobackground">
        <!-- The content of this hgroup is replaced programmatically through the slide_config.json. -->
    <hgroup class="auto-fadein">
      <h1 data-config-title><!-- populated from slide_config.json --></h1>
      <h2 data-config-subtitle><!-- populated from slide_config.json --></h2>
      <p data-config-presenter><!-- populated from slide_config.json --></p>
            <p style="margin-top: 6px; margin-left: -2px;">2020-05-27</p>
          </hgroup>
  </slide>

<slide class=""><hgroup><h2>Reference</h2></hgroup><article  class="smaller " id="reference">

<p>Pavkovic, M., Pantano, L., Gerlach, C.V. et al. Multi omics analysis of fibrotic kidneys in two mouse models. Sci Data 6, 92 (2019).</p>

<ul>
<li><a href='https://doi.org/10.1038/s41597-019-0095-5' title=''>https://doi.org/10.1038/s41597-019-0095-5</a></li>
<li><a href='https://www.nature.com/articles/s41597-019-0095-5#citeas' title=''>https://www.nature.com/articles/s41597-019-0095-5#citeas</a></li>
<li>Mouse fibrotic kidney browser: <a href='http://hbcreports.med.harvard.edu/fmm/' title=''>http://hbcreports.med.harvard.edu/fmm/</a></li>
<li>Data on Zenodo: <a href='https://zenodo.org/record/2592516' title=''>https://zenodo.org/record/2592516</a></li>
</ul>

</article></slide><slide class=""><hgroup><h2>Data sources</h2></hgroup><article  class="smaller " id="data-sources">

<table class = 'rmdtable'>
<col width="52.631579%" />
<col width="47.368421%" />
<tr class="header">
<th align="left">Doc</th>
<th align="left">URL</th>
</tr>
<tr class="odd">
<td align="left">Total RNA for the experiment on Unilateral ureter obstruction (UUO) model</td>
<td align="left"><a href='https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118339' title=''>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118339</a></td>
</tr>
</table>

</article></slide><slide class=""><hgroup><h2>Parameters</h2></hgroup><article  class="smaller " id="parameters">

<pre class = 'prettyprint lang-r'>#### Define parameters for the analysis ####

## Keep a trace of the original parameters
par.ori &lt;- par(no.readonly = TRUE)

## Analysis parameters
parameters &lt;- list(
  epsilon = 0.1,
  minCount = 10,
  forceDownload = FALSE)

kable(as.data.frame(parameters))</pre>

<table class = 'rmdtable'>
<tr class="header">
<th align="right">epsilon</th>
<th align="right">minCount</th>
<th align="left">forceDownload</th>
</tr>
<tr class="odd">
<td align="right">0.1</td>
<td align="right">10</td>
<td align="left">FALSE</td>
</tr>
</table>

</article></slide><slide class=""><hgroup><h2>Output directories</h2></hgroup><article  class="smaller " id="output-directories">

<pre class = 'prettyprint lang-r'>#### Output directories ####
outdirs &lt;- list()
outdirs$main &lt;- getwd()

## Data directory, where the data will be downloaded and uncompressed
outdirs$data &lt;- file.path(outdirs$main, &quot;data&quot;)
dir.create(outdirs$data, recursive = TRUE, showWarnings = FALSE)

## Main result directory
outdirs$results &lt;- file.path(outdirs$main, &quot;results&quot;)

# Transcriptome results
outdirs$transcriptome &lt;- file.path(outdirs$main, &quot;results/transcriptome&quot;)
dir.create(outdirs$transcriptome, recursive = TRUE, showWarnings = FALSE)</pre>

</article></slide><slide class=""><hgroup><h2>Download transcriptome data</h2></hgroup><article  class="smaller " id="download-transcriptome-data">

<pre class = 'prettyprint lang-r'>#### Download transcriptome data ####
archiveFile &lt;- &quot;MouseKidneyFibrOmics-v1.0.zip&quot;
archiveURL &lt;- file.path(&quot;https://zenodo.org/record/2592516/files/hbc&quot;, archiveFile)
localDataArchive &lt;- file.path(outdirs$data, archiveFile)

if (file.exists(localDataArchive) &amp; !parameters$forceDownload) {
  message(&quot;Data archive already downloaded:\n\t&quot;, localDataArchive)
} else {
  message(&quot;Downloading data archive from zenodo: &quot;, archiveURL)
  download.file(url = archiveURL, destfile =  localDataArchive)
  
  ## Uncompess the archive
  message(&quot;Uncompressing data archive&quot;)
  unzip(zipfile = localDataArchive, exdir = outdirs$data)
}

## Define destination directory
# outdirs$csv &lt;- file.path(outdirs$data, &quot;CSV&quot;)
# dir.create(outdirs$csv, showWarnings = FALSE, recursive = TRUE)</pre>

</article></slide><slide class=""><hgroup><h2>Load raw counts</h2></hgroup><article  class="smaller " id="load-raw-counts">

<pre class = 'prettyprint lang-r'>#### Load raw counts data table ####
rawCountFile &lt;- file.path(
  outdirs$data,
  &quot;hbc-MouseKidneyFibrOmics-a39e55a/tables/uuo/results/counts/raw_counts.csv.gz&quot;)
rawCounts &lt;- read.csv(file = rawCountFile, header = 1, row.names = 1)</pre>

<p>The RNA-seq trnascriptome data was loaded as raw counts. This table contains 46679 rows (genes) and 15 columns (samples).</p>

</article></slide><slide class=""><hgroup><h2>Build metadata table</h2></hgroup><article  class="smaller " id="build-metadata-table">

<pre class = 'prettyprint lang-r'>#### Build metadata table ####
metadata &lt;- data.frame(sampleName = colnames(rawCounts))
metadata[ , c(&quot;condition&quot;, &quot;sampleNumber&quot;)] &lt;- 
  str_split_fixed(string = metadata$sampleName, pattern = &quot;_&quot;, n = 2)

## Colors per condition
colPerCondition &lt;- c(normal = &quot;#BBFFBB&quot;,
                     day3 = &quot;#FFFFDD&quot;, 
                     day7 = &quot;#FFDD88&quot;,
                     day14 = &quot;#FF4400&quot;)
metadata$color &lt;- colPerCondition[metadata$condition]</pre>

</article></slide><slide class=""><hgroup><h2>Distribution of raw counts</h2></hgroup><article  class="smaller " id="distribution-of-raw-counts">

<pre class = 'prettyprint lang-r'>hist(unlist(rawCounts), breaks = 1000, 
     main = &quot;Raw count distribution&quot;, 
     xlab = &quot;Raw counts&quot;, 
     ylab = &quot;Number of genes (all samples)&quot;)</pre>

<div class="figure" style="text-align: center">
<img src="figures/mouse-kidney_RNA-seq_count_distrib-1.png" alt="Distribution of raw counts" width="70%" />

<p class="caption">

Distribution of raw counts

</p></div>

<p>The distribution of raw counts is not very informative, because the range is defined by some outlier, i.e. a gene having a huge number of reads. Even with strong zoom on the abcsissa range from 0 to 500, the histogram shows a steep drop in the first bins.</p>

</article></slide><slide class=""><hgroup><h2>Distribution of raw counts - truncated abscissa</h2></hgroup><article  class="smaller " id="distribution-of-raw-counts---truncated-abscissa">

<pre class = 'prettyprint lang-r'>#### Count distrib - truncated abscissa ####
hist(unlist(rawCounts), breaks = 500000, 
     main = &quot;Raw count distribution&quot;, 
     xlab = &quot;Raw counts (truncated abscissa&quot;, 
     ylab = &quot;Number of genes (all samples)&quot;,
     xlim = c(0, 500))</pre>

<div class="figure" style="text-align: center">
<img src="figures/mouse-kidney_RNA-seq_count_distrib_truncated-1.png" alt="Distribution of raw counts" width="70%" />

<p class="caption">

Distribution of raw counts

</p></div>

</article></slide><slide class=""><hgroup><h2>Log2-transformed counts</h2></hgroup><article  class="smaller " id="log2-transformed-counts">

<p>A typical approach is to normalise the counts by applying a log2 transformation . This however creates a problem when the counts of a given gene in a given sample is 0. To circumvent this, we can add an epsilon (\(\epsilon = 0.1\)) before the log2 transformation.</p>

<pre class = 'prettyprint lang-r'>#### Log2 transformatiojn of the counts ####
log2Counts &lt;- log2(rawCounts + parameters$epsilon)
hist(unlist(log2Counts), breaks = seq(from = -5, to = 22, by = 0.1),
     main = &quot;log2-counts distribution&quot;,
     xlab = &quot;log2(Counts + epsilon)&quot;, 
     ylab = &quot;Number of genes&quot;, col = &quot;#BBFFBB&quot;)</pre>

<p><img src="figures/mouse-kidney_log2_counts-1.png" width="70%" style="display: block; margin: auto;" /></p>

</article></slide><slide class=""><hgroup><h2>Two-columns test</h2></hgroup><article  class="smaller " id="two-columns-test">

<div class="columns">
<div class="column">
<p>contents 1 …</p></div>

<div class="column">
<p>pas l’air de marcher …</p></div></div>

</article></slide><slide class=""><hgroup><h2>Box plots</h2></hgroup><article  class="smaller " id="box-plots">

<p>We can now inspect the distribution of counts per sample with the <code>boxplot()</code> function.</p>

<pre class = 'prettyprint lang-r'>#### Box plots ####
par(mar = c(4, 6, 5, 1))
boxplot(log2Counts, 
        col = metadata$color,
        horizontal = TRUE, 
        las = 1, 
        main = &quot;log2-transformed&quot;, 
        xlab = &quot;log2(counts)&quot;)</pre>

<p><img src="figures/mouse-kidney_box%20plots-1.png" width="25%" style="display: block; margin: auto;" /></p>

<pre class = 'prettyprint lang-r'>par(par.ori)</pre>

</article></slide><slide class=""><hgroup><h2>Box plot comment</h2></hgroup><article  class="smaller " id="box-plot-comment">

<p>We notice an obvious problem: the vast majority of counts is very small. This can result from different causes, which will not be investigated in this context.</p>

</article></slide><slide class=""><hgroup><h2>Gene filtering</h2></hgroup><article  class="smaller " id="gene-filtering">

<pre class = 'prettyprint lang-r'>## Filter out the genes with very low counts in all conditions
undetectedGenes &lt;- apply(rawCounts, MARGIN = 1, FUN = sum) &lt; parameters$minCount
# table(undetectedGenes)
log2CountsFiltered &lt;- log2Counts[!undetectedGenes, ]</pre>

<p>We filtered out all the genes whose maximal count value across all samples was lower than 10. Among the 46679 genes from the raw count table, 20851 were considered undetected according to this criterion. We use the remaining 25828 genes for the subsequent analyses.</p>

</article></slide><slide class=""><hgroup><h2>Gene filtered boxplot</h2></hgroup><article  class="smaller " id="gene-filtered-boxplot">

<pre class = 'prettyprint lang-r'>#### Box plots after filtering ####
par(mar = c(4, 6, 5, 1))
boxplot(log2CountsFiltered, 
        col = metadata$color,
        horizontal = TRUE, 
        las = 1,
        main = &quot;Filtered genes&quot;, 
        xlab = &quot;log2(counts)&quot;)</pre>

<p><img src="figures/mouse-kidney_boxplot_filtered-1.png" width="32%" style="display: block; margin: auto;" /></p>

<pre class = 'prettyprint lang-r'>par(par.ori)</pre>

</article></slide><slide class=""><hgroup><h2>Normalisation (more precisely: scaling)</h2></hgroup><article  class="smaller " id="normalisation-more-precisely-scaling">

<p>Before going any further, it is important to ensure some normalisation of the counts, in order to correct for biases due to inter-sample differences in sequencing depth.</p>

<p>For the sake of simplicity, we will use here a very simple criterion: median-based normalisation. The principle is to multiply the counts of each sample by a scaling factor in order to bring each sample to the same median count.</p>

</article></slide><slide class=""><hgroup><h2>Normalisation code</h2></hgroup><article  class="smaller " id="normalisation-code">

<pre class = 'prettyprint lang-r'>#### Median-based normalisation ####
sampleMedians &lt;- apply(log2CountsFiltered, 2, median)
seriesMedian &lt;- median(sampleMedians)
scalingFactors &lt;- seriesMedian / sampleMedians

log2CountsStandardised &lt;- data.frame(matrix(
  nrow = nrow(log2CountsFiltered),
  ncol = ncol(log2CountsFiltered)))
colnames(log2CountsStandardised) &lt;- colnames(log2CountsFiltered)
rownames(log2CountsStandardised) &lt;- rownames(log2CountsFiltered)
for (j in 1:ncol(log2CountsFiltered)) {
  log2CountsStandardised[, j] &lt;- log2CountsFiltered[, j] * scalingFactors[j]
}

## Check the remaining medians
apply(log2CountsStandardised, 2, median)</pre>

<pre >day14_12 day14_13 day14_14 day14_15   day3_4   day3_5   day3_6   day3_7  day7_10  day7_11   day7_8   day7_9 normal_1 normal_2 normal_3 
6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 6.772065 </pre>

</article></slide><slide class=""><hgroup><h2>Normalization boxplot</h2></hgroup><article  class="smaller " id="normalization-boxplot">

<pre class = 'prettyprint lang-r'>#### Box plots after scaling ####
par(mar = c(4, 6, 5, 1))
boxplot(log2CountsStandardised, 
        col = metadata$color,
        horizontal = TRUE, 
        las = 1,
        main = &quot;Median-based scaled&quot;, 
        xlab = &quot;log2(counts&quot;)</pre>

<p><img src="figures/mouse-kidney_median_normalisation_boxplot-1.png" width="32%" style="display: block; margin: auto;" /></p>

<pre class = 'prettyprint lang-r'>par(par.ori)</pre>

</article></slide><slide class=""><hgroup><h2>Session info</h2></hgroup><article  class="smaller " id="session-info">

<pre class = 'prettyprint lang-r'>sessionInfo()</pre>

<pre >R version 3.6.3 (2020-02-29)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /shared/mfs/data/software/miniconda/envs/r-3.6.3/lib/libopenblasp-r0.3.9.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] stringr_1.4.0 knitr_1.28   

loaded via a namespace (and not attached):
 [1] compiler_3.6.3  magrittr_1.5    tools_3.6.3     htmltools_0.4.0 yaml_2.2.1      Rcpp_1.0.4.6    stringi_1.4.6   rmarkdown_2.1   highr_0.8       xfun_0.14       digest_0.6.25   rlang_0.4.6     evaluate_0.14  </pre></article></slide>


  <slide class="backdrop"></slide>

</slides>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!-- map slide visiblity events into shiny -->
<script>
  (function() {
    if (window.jQuery) {
       window.jQuery(document).on('slideleave', function(e) {
         window.jQuery(e.target).trigger('hidden');
      });
       window.jQuery(document).on('slideenter', function(e) {
         window.jQuery(e.target).trigger('shown');
      });
    }
  })();
</script>

</body>
</html>
